.github/workflows/cargo-deny-pr.yml

```diff
@@ -0,0 +1,24 @@
+name: cargo deny
+on:
+  push:
+    paths:
+      - '**/Cargo.toml'
+      - '**/Cargo.lock'
```

Comment:
We don't have this checked into Git so I think it doesn't do anything?

Reply:
Not at the moment no, but we should put lock into Git after this so it can be used to quick screen every new update that comes through via dependabot and also that means that you can easily continuously test all updates what the users are actually using which also means that the --locked version is tested. It also is useful in changelog to see if e.g. a library that was implementation detail that went up a SemVer or especially non-implementation detail which might affect pub APIs.

Reply:
We are a library though so I am not sure what benefit we are getting from checking in `Cargo.lock`?

On the contrary, I'd say it is better to not pin specific versions. A new user will get the same dependencies as our CI runs if they add `rust-libp2p` to their `Cargo.toml`, meaning in case the builds breaks due to a bad patch semver upgrade f.e., we should know earlier than if we were to check in a `Cargo.lock`.

Upgrades outside of the allowed semver range are performed by dependabot and if they contain publicly exposed types, we already note these in our changelog. (At the same time, we try to limit these cases to avoid unnecessary breaking changes for our users.)

To get back on topic, I think we should remove these triggers because they don't do anything for us and are effectively dead configuration.

Reply:
Cargo.lock ensures that you test with all locked versions - as a library - and be aware of changes from downstream that the library still works for anyone given Cargo.lock - think your Cargo.lock as some manifest that you've tested and you're actively supporting.

It also ensures that you get security advisories that are triggered from Cargo.lock and not from Cargo.toml which may not reflect the full reality of the usage - .lock also shows all the conflciting versions e.g. if you use time 0.1 due to using default feature-set f.ex. that would be easy to fix once it's visible.

Reply:
I am not quite sure I follow.

Assume a security advisory in some crate `foo` that is vulnerable in version: `<= 0.1.10` and `rust-libp2p` specifies `0.1.10`. Let there be a security fix in `0.1.11`. With the requirement `0.1.10`, we only express that we need _at least_ `0.1.10` but any greater patch version is fine too. Thus any binary can freely update to `0.1.11` on their end. I don't see how `rust-libp2p` is vulnerable here or needs any change to its manifest. Any security advisory issue opened on our project is basically a "can't fix" situation.

Now, in the time between the advisory being published and the fix being available, `cargo audit` will still open an issue on our repository but there isn't a whole lot we can do. I don't see how checking in a `Cargo.lock` file improves anything in this situation. Instead of not having to do anything, we need to explicitly bump the version in a PR that will likely only touch the lockfile: something that is irrelevant for our users.

On a more general note: I don't think it is good practice to only support a locked range of versions. On the contrary, we should strive to be compatible with any version combination that is possible given our `Cargo.toml`. I don't see how a versioned lockfile is helping us with that.

Reply:
.lock would make sure that you always test with locked set of dependencies and ensures that every time there is transitive dependency update you know that your software still works.. instead of users reporting it - despite this not leaving with you any control at least you are aware of it and can possibility mitigate e.g. if crate did a breaking change where they didn't bump semver - this happens quite a bit ...

Reply:
I basically - even as a library - screen / grate all .lock bumps to make sure everything still works. It's additional chore but imo worth it when I just do @dependabot merge for all of them

Reply:
Anyways I addressed your .lock out so can always add that back if you choose to add Cargo.lock

Reply:
Thanks for the input! Definitely something to consider for the future. I'll open an issue to track it as a separate discussion!